`e2e` failing the abci grammar test

## Bug Report

### Setup

**CometBFT version** (use `cometbft version` or `git rev-parse --verify HEAD` if installed from source): #2454 (base branch: `feature/pbts`, a feature branch)

**Have you tried the latest version**: yes

**ABCI app** (name for built-in, URL for self-written if it's publicly available): `e2e` app

**Environment**:
- **OS** (e.g. from /etc/os-release): Linux (docker)
- **Install tools**:
- **Others**:

**node command runtime flags**: e2e defaults

### Config

According to manifest file `test/e2e/networks/ci.toml`

### What happened?

The test case that checks the nodes' traces against the ABCI grammar is consistently failing on `validator05` when we introduce clock skew in `validator01`

### What did you expect to happen?

The test should pass

### How to reproduce it

Run `e2e` tests using `test/e2e/networks/ci.toml` on #2454

### Logs

```
--- FAIL: TestCheckABCIGrammar (3.13s)
    --- FAIL: TestCheckABCIGrammar/validator05 (0.34s)
        e2e_test.go:63:
                Error Trace:    /Users/sergio/work/scm/github/comet1/test/e2e/tests/abci_test.go:24
                                                        /Users/sergio/work/scm/github/comet1/test/e2e/tests/e2e_test.go:63
                Error:          Received unexpected error:
                                The error: "Invalid execution: parser was expecting one of [finalize_block,prepare_proposal,process_proposal], got [$] instead." has occurred at height 0.

                                Full execution:
                                0: process_proposal
                Test:           TestCheckABCIGrammar/validator05
```

... and enhanced log:

```
--- FAIL: TestCheckABCIGrammar (2.73s)
    --- FAIL: TestCheckABCIGrammar/validator05 (0.34s)
        e2e_test.go:63:
                Error Trace:    /Users/sergio/work/scm/github/comet1/test/e2e/tests/abci_test.go:24
                                                        /Users/sergio/work/scm/github/comet1/test/e2e/tests/e2e_test.go:63
                Error:          Received unexpected error:
                                The error: "Invalid execution: parser was expecting one of [finalize_block,prepare_proposal,process_proposal], got [$] instead." has occurred at height 0.

                                Full execution:
                                0: info prepare_proposal process_proposal check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx process_proposal check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx check_tx
                Test:           TestCheckABCIGrammar/validator05
```
### `dump_consensus_state` output

N/A

### Anything else we need to know

The root cause has already been identified by @nenadmilosevic95 and a fix exists.

--- a/test/e2e/pkg/grammar/checker.go
+++ b/test/e2e/pkg/grammar/checker.go
@@ -91,7 +91,7 @@ func (g *Checker) filterLastHeight(reqs []*abci.Request) ([]*abci.Request, int) 
 	pos := len(reqs) - 1
 	cnt := 0
 	// Find the last commit.
-	for pos > 0 && g.getRequestTerminal(reqs[pos]) != Commit {
+	for pos >= 0 && g.getRequestTerminal(reqs[pos]) != Commit {
 		pos--
 		cnt++
 	}
@@ -129,6 +129,7 @@ func (g *Checker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) 
 	if len(reqs) == 0 {
 		return false, errors.New("execution with no ABCI calls")
 	}
+	fullExecution := g.getExecutionString(reqs)
 	r := g.filterRequests(reqs)
 	// Check if the execution is incomplete.
 	if len(r) == 0 {
@@ -139,7 +140,7 @@ func (g *Checker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) 
 	if errors == nil {
 		return true, nil
 	}
-	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(execution))
+	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(fullExecution))
 }
 
 // verifyCleanStart verifies if a specific execution is a valid execution.

--- a/test/e2e/pkg/grammar/checker_test.go
+++ b/test/e2e/pkg/grammar/checker_test.go
@@ -34,9 +34,7 @@ type ABCIExecution struct {
 var consExecPart = []ABCIExecution{
 	// consensus-height = finalizeBlock commit
 	{[]*abci.Request{finalizeBlock, commit}, true},
-	{[]*abci.Request{}, false},
 	{[]*abci.Request{commit}, false},
-	{[]*abci.Request{finalizeBlock}, false},
 	// consensus-height = *consensus-round finalizeBlock commit
 	// consensus-height = *consensus-round finalizeBlock commit
 	// consensus-round = proposer
@@ -119,8 +117,8 @@ func TestVerifyCleanStart(t *testing.T) {
 		// state-sync = success-sync
 		{[]*abci.Request{offerSnapshot, applyChunk}, true},
 		{[]*abci.Request{offerSnapshot, applyChunk, applyChunk}, true},
-		{[]*abci.Request{applyChunk, finalizeBlock, commit}, false},
-		{[]*abci.Request{offerSnapshot, finalizeBlock, commit}, false},
+		{[]*abci.Request{applyChunk}, false},
+		{[]*abci.Request{offerSnapshot}, false},
 		// state-sync = *state-sync-attempt success-sync
 		{[]*abci.Request{offerSnapshot, applyChunk, offerSnapshot, applyChunk}, true},
 		{[]*abci.Request{offerSnapshot, applyChunk, applyChunk, applyChunk, offerSnapshot, applyChunk}, true},
@@ -174,13 +172,17 @@ func TestVerifyRecovery(t *testing.T) {
 }
 
 func TestFilterLastHeight(t *testing.T) {
-	reqs := []*abci.Request{initChain, finalizeBlock, commit}
+	reqs := []*abci.Request{initChain, finalizeBlock}
 	checker := NewGrammarChecker(DefaultConfig())
-	rr, n := checker.filterLastHeight(reqs)
-	require.Equal(t, len(reqs), len(rr))
+	r, n := checker.filterLastHeight(reqs)
+	require.Equal(t, len(r), 0)
+	require.Equal(t, n, 2)
+	reqs = append(reqs, commit)
+	r, n = checker.filterLastHeight(reqs)
+	require.Equal(t, len(r), len(reqs))
 	require.Zero(t, n)
-	reqs = append(reqs, finalizeBlock)
-	rrr, n := checker.filterLastHeight(reqs)
-	require.Equal(t, len(rr), len(rrr))
-	require.Equal(t, 1, n)
+	reqs = append(reqs, []*abci.Request{prepareProposal, processProposal}...)
+	r, n = checker.filterLastHeight(reqs)
+	require.Equal(t, len(r), 3)
+	require.Equal(t, n, 2)
 }

